gufe/base.py

```diff
@@ -0,0 +1,381 @@
+    dct.update(module_qualname(obj))
+    return dct
+
+
+def dict_encode_dependencies(obj: GufeTokenizable) -> Dict:
+    return modify_dependencies(
+        obj.to_shallow_dict(),
+        to_dict,
+        is_gufe_obj,
+        top=True
+    )
+
+
+def key_encode_dependencies(obj: GufeTokenizable) -> Dict:
+    return modify_dependencies(
+        obj.to_shallow_dict(),
+        lambda obj: obj.key.to_dict(),
+        is_gufe_obj,
+        top=True
+    )
+
+
+# decode options
+def from_dict(dct) -> GufeTokenizable:
+    dct = copy.deepcopy(dct)
+
+    for key, val in dct.items():
+        if is_gufe_dict(val):
+            dct[key] = from_dict(val)
+
+    obj = _from_dict(dct)
+    try:
+        thing = TOKENIZABLE_REGISTRY[obj.key]()
+        # weakref will return None if the object was deleted
+        if thing is None:
+            return obj
+        else:
+            return thing
+    except KeyError:
+        return obj
+
+
+def _from_dict(dct: Dict) -> GufeTokenizable:
+    module = dct.pop('__module__')
+    qualname = dct.pop('__qualname__')
+    if (qualname is None) or (module is None):
+        raise KeyError("`__qualname__` or `__module__` key not found; unable "
+                       "to reconstitute from dict")
+
+    cls = get_class(module, qualname)
+    return cls._from_dict(dct)
+
+
+def dict_decode_dependencies(dct: Dict) -> GufeTokenizable:
+    return from_dict(
+        modify_dependencies(dct, from_dict, is_gufe_dict, top=True)
+    )
+
+
+def key_decode_dependencies(dct: Dict) -> GufeTokenizable:
+    # this version requires that all dependent objects are already registered
+    # responsibility of the storage system that uses this to do so
+    dct = modify_dependencies(
+        dct,
+        lambda d: TOKENIZABLE_REGISTRY[GufeKey(d[":gufe-key:"])](),
+        is_gufe_key_dict,
+        top=True
+    )
+    return from_dict(dct)
+
+
+# FROM dask.base
+# Pass `usedforsecurity=False` for Python 3.9+ to support FIPS builds of Python
+_PY_VERSION = parse_version(".".join(map(str, sys.version_info[:3])))
+_md5: Callable
+if _PY_VERSION >= parse_version("3.9"):
+
+    def _md5(x, _hashlib_md5=hashlib.md5):
+        return _hashlib_md5(x, usedforsecurity=False)
+else:
+    _md5 = hashlib.md5
+
+
+def normalize_object(o: GufeTokenizable):
+    try:
+        return o._gufe_tokenize()
+    except AttributeError:
+        raise ValueError("Cannot normalize object without `_gufe_tokenize` "
+                         "method.")
+
+
+def tokenize(obj: GufeTokenizable) -> str:
+    """Generate a deterministic, relatively-stable token from a
+    `GufeTokenizable` object.
+
+    Examples
+    --------
+    >>> from gufe import SolventComponent
+    >>> s = SolventComponent()
+    >>> tokenize(s)
+    'e6eef7519854d35a5ce6c84136b3684c'
+
+    >>> tokenize(s) == tokenize(SolventComponent.from_dict(s.to_dict()))
+    True
+
+    """
+    hasher = _md5(str(normalize_object(obj)).encode())
+    return hasher.hexdigest()
```

Comment:
As stripped down as we've gotten it, I don't think we need all the dask-like structure here. 
* `_md5` isn't really reused in multiple places, so doesn't need to be module-level. Plus, I thought we were dropping Python 3.8, which means that we don't even need the branching to select it.
* `normalize_object` just says "I'll give you the normalized dict if you give me a gufe object, otherwise error." Do we have a use case where that error is useful? Can't this be removed and just use `_gufe_tokenize` when needed?
* `tokenize` takes the normalized dict and makes a hex token out of it. Can this just be inlined into  the `key` property?

Reply:
Agree, simplifying now. Can re-add complexity later if needed.

Reply:
I do like keeping the `tokenize` standalone function, even if its use is internal, so have not inlined it into `GufeTokenizable.key`.